copy/copy.go

```diff
@@ -1390,7 +1414,7 @@ func (c *copier) copyBlobFromStream(ctx context.Context, srcStream io.Reader, sr
 		}
```

Comment:
Oops:
Above, the
```go
	// If we can modify the layer's blob, set the desired algorithm for it to be set in the manifest.
	if canModifyBlob && !isConfig {
		uploadedInfo.CompressionAlgorithm = &desiredCompressionFormat
	}
```
condition is no longer a no-op if `CompressionOperation != types.Compress` (so just a simple copy of a `zstd` or uncompressed data is now changed to a `gzip` MIME type AFAICS).

I think replacing that condition with `uploadCompressorName != internalblobinfocache.UnknownCompression && uploadCompressorName != internalblobinfocache.Uncompressed` would work, but it feels a bit that there should be a more elegant way.

Reply:
One possible approach is that the decision made has been indeed tracked in `uploadCompressorName`, so this could be
```go
if uploadCompressorName != internalblobinfocache.UnknownCompression && uploadCompressorName != internalblobinfocache.Uncompressed {
    algo, err := compression.AlgorithmByName(uploadCompressorName);
    if err { /* ??! */ }
    uploadedInfo.CompressionAlgorithm = &algo
}
```
which is _so_ ugly, but we can’t nicely and succinctly express `uploadCompressor = enum { unknown, uncompressed, algorithm(compression.Algorithm) }` in Go. Alternatively, the `if`/`else if` cascade could set yet another variable with the right `uploadedCompressionAlgorithm` value, risking inconsistency with `uploadCompressorName`.

So here’s hoping you have a much more elegant idea.

Reply:
I suppose we could initialize `desiredCompressionFormat` to `&c.compressionFormat`, and clear or change its value in the "Deal with layer compression/decompression if necessary" sections where we update `uploadCompressorName` with values that would conflict with it.  Then we could assign it to `uploadedInfo.CompressionAlgorithm` unconditionally.  The final `else` clause would still need to check `isCompressed` to decide between `&compressionFormat` and `nil`.  The additional diff looks more or less like https://gist.github.com/nalind/3e12d2be6f6858720851739de6a5d6ea.

Reply:
Works for me.
- `s/desiredCompressionFormat/uploadCompressionFormat/` , then, I guess — it now records the outcome, not the input
- in the `!canModify` case, set it always to `nil` to unambiguously trigger no manifest edits.

Reply:
Okay, added the gist patch, made these changes, and rebased.